google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java

```diff
@@ -0,0 +1,287 @@
+      this.name = name;
+    }
+
+    static SessionId of(String name) {
+      Map<String, String> parts = NAME_TEMPLATE.match(name);
+      Preconditions.checkArgument(
+          parts != null, "Name should conform to pattern %s: %s", NAME_TEMPLATE, name);
+      return of(
+          parts.get("project"), parts.get("instance"), parts.get("database"), parts.get("session"));
+    }
+
+    /** Creates a {@code SessionId} given project, instance, database and session IDs. */
+    static SessionId of(String project, String instance, String database, String session) {
+      return new SessionId(new DatabaseId(new InstanceId(project, instance), database), session);
+    }
+
+    DatabaseId getDatabaseId() {
+      return db;
+    }
+
+    String getName() {
+      return name;
+    }
+  }
+
+  /**
+   * Encapsulates state to be passed to the {@link SpannerRpc} layer for a given session. Currently
+   * used to select the {@link io.grpc.Channel} to be used in issuing the RPCs in a Session.
+   */
+  static class SessionOption {
+    private final SpannerRpc.Option rpcOption;
+    private final Object value;
+
+    SessionOption(SpannerRpc.Option option, Object value) {
+      this.rpcOption = checkNotNull(option);
+      this.value = value;
+    }
+
+    static SessionOption channelHint(long hint) {
+      return new SessionOption(SpannerRpc.Option.CHANNEL_HINT, hint);
+    }
+
+    SpannerRpc.Option rpcOption() {
+      return rpcOption;
+    }
+
+    Object value() {
+      return value;
+    }
+  }
+
+  static Map<SpannerRpc.Option, ?> optionMap(SessionOption... options) {
+    if (options.length == 0) {
+      return Collections.emptyMap();
+    }
+    Map<SpannerRpc.Option, Object> tmp = Maps.newEnumMap(SpannerRpc.Option.class);
+    for (SessionOption option : options) {
+      Object prev = tmp.put(option.rpcOption(), option.value());
+      checkArgument(prev == null, "Duplicate option %s", option.rpcOption());
+    }
+    return ImmutableMap.copyOf(tmp);
+  }
+
+  private final class BatchCreateSessionsRunnable implements Runnable {
+    private final long channelHint;
+    private final int sessionCount;
+    private final SessionConsumer consumer;
+
+    private BatchCreateSessionsRunnable(
+        int sessionCount, long channelHint, SessionConsumer consumer) {
+      Preconditions.checkNotNull(consumer);
+      Preconditions.checkArgument(sessionCount > 0, "sessionCount must be > 0");
+      this.channelHint = channelHint;
+      this.sessionCount = sessionCount;
+      this.consumer = consumer;
+    }
+
+    @Override
+    public void run() {
+      List<SessionImpl> sessions = null;
+      int remainingSessionsToCreate = sessionCount;
+      while (remainingSessionsToCreate > 0) {
+        try {
+          sessions = internalBatchCreateSessions(remainingSessionsToCreate, channelHint);
+        } catch (Throwable t) {
+          consumer.onSessionCreateFailure(t, remainingSessionsToCreate);
+          break;
+        }
```

Comment:
would we not retry  ? 
If retries are  already handled by gax layer, could u verify what error codes are retried ?

Reply:
Gax handles the retries, and will do so based on the settings here: https://github.com/googleapis/googleapis/blob/6b2ba2ae3124c22ecb56af7102c78110b8576671/google/spanner/v1/spanner_gapic.yaml#L93

That means that the method will retry on:
* Server returns error code Unavailable
* Deadline_Exceeded (client side) after 60 seconds, and it will give up retrying after 600 seconds.